File: frontend/src/app/model/CustomResponse.ts
export class CustomResponse<T> {
  private readonly _data: T;
  private readonly _error: string;

  constructor(data: T, error: string) {
    this._data = data;
    this._error = error;
  }

  get data(): T | null {
    return this._data;
  }

  get error(): string | null {
    return this._error;
  }
}

File: backend/src/main/java/de/alex/blogster_rest_api/http/model/response/CustomResponse.java
package de.alex.blogster_rest_api.http.model.response;

public class CustomResponse<T> {
    private final T data;
    private final String error;

    public CustomResponse(T data) {
        this.data = data;
        this.error = null;
    }

    public CustomResponse(String error) {
        this.data = null;
        this.error = error;
    }

    public T getData() {
        return data;
    }

    public String getError() {
        return error;
    }
}
